wrapper/src/test/java/integration/refactored/container/tests/HikariTests.java

```diff
@@ -110,4 +119,69 @@ public void testOpenConnectionWithDataSourceClassName() throws SQLException {
     assertTrue(conn.isValid(10));
     conn.close();
   }
+
+  @TestTemplate
+  public void testServerFailoverWithIdleHikariConnections() throws SQLException, InterruptedException {
+    HikariDataSource ds = new HikariDataSource();
+    ds.setDataSourceClassName(AwsWrapperDataSource.class.getName());
+
+    // Configure the connection pool:
+    ds.setUsername(TestEnvironment.getCurrent().getInfo().getDatabaseInfo().getUsername());
+    ds.setPassword(TestEnvironment.getCurrent().getInfo().getDatabaseInfo().getPassword());
+    ds.setMaximumPoolSize(POOL_SIZE);
+    ds.setMinimumIdle(POOL_SIZE);
+
+    // Configure AwsWrapperDataSource:
+    ds.addDataSourceProperty(
+        "jdbcProtocol", DriverHelper.getDriverProtocol());
+    ds.addDataSourceProperty("databasePropertyName", "databaseName");
+    ds.addDataSourceProperty("portPropertyName", "port");
+    ds.addDataSourceProperty("serverPropertyName", "serverName");
+
+    // Specify the driver-specific DataSource for AwsWrapperDataSource:
+    ds.addDataSourceProperty("targetDataSourceClassName", DriverHelper.getDataSourceClassname());
+    // Configuring driver-specific DataSource:
+    Properties targetDataSourceProps = new Properties();
+    targetDataSourceProps.setProperty(
+        "serverName",
+        TestEnvironment.getCurrent()
+            .getInfo()
+            .getDatabaseInfo()
+            .getClusterEndpoint());
+    targetDataSourceProps.setProperty(
+        "databaseName",
+        TestEnvironment.getCurrent().getInfo().getDatabaseInfo().getDefaultDbName());
+    targetDataSourceProps.setProperty("url", ConnectionStringHelper.getClusterUrl());
+
+    targetDataSourceProps.setProperty(PropertyDefinition.PLUGINS.name, "auroraConnectionTracker,failover");
+    ds.addDataSourceProperty("targetDataSourceProperties", targetDataSourceProps);
+
+    auroraUtil.queryInstanceId(
+        TestEnvironment.getCurrent().getInfo().getRequest().getDatabaseEngine(),
+        ds.getConnection());
+
+    // Crash Instance1 and nominate a new writer
+    auroraUtil.failoverClusterAndWaitUntilWriterChanged(TestEnvironment.getCurrent()
+        .getInfo()
+        .getDatabaseInfo()
+        .getInstances()
+        .get(0)
+        .getInstanceName());
+
+    // Assert failover has occurred.
+    SQLException exception =
+        assertThrows(
+            SQLException.class,
+            () ->
+                auroraUtil.queryInstanceId(
+                    TestEnvironment.getCurrent().getInfo().getRequest().getDatabaseEngine(),
+                    ds.getConnection()));
+    assertEquals("08S02", exception.getSQLState());
+
+    // Ensure that all idle connections are closed.
+    // HikariCP will throw a SQLTransientConnectionException if the connection is closed or invalid.
+    for (int i = 0; i < POOL_SIZE; i++) {
+      assertThrows(SQLTransientConnectionException.class, () -> ds.getConnection().isValid(3));
```

Comment:
Hikari can open new connection when calling ds.getConnection() and that may ruin the purpose of the test. We need to come up to another solution to check connections that are already in the pool and make sure there's no valid (i.e. not yet closed) connections to nodes from line 159 and line 170.

Reply:
Should we check isClosed() instead?

Reply:
Both isClosed and isValid returns the same result. The exception is thrown by Hikari before any database execution